lib/io.js

```diff
@@ -649,35 +693,110 @@
       return aroundClusterSwitch(function() {
         var lazyConns = toLazyAltConnections(targetCluster, transport);
         return findFirstConnection(lazyConns, getManualMainRouter);
+      }, function () {
+        logger.error('Unable to switch to cluster %s', targetCluster.name);
+        return Promise.resolve(false);
       })
     }
 
     function mainConnection(conn) {
-      return conn.connect();
+      if (o.authOpts().enabled) {
+        return conn.connect().then(function () {
+          return authMech(o, conn, o.getTopologyId());
+        }).then(function (authMechs) {
+          logger.tracef(authMechs);
+          if (!_.contains(authMechs, o.authOpts().saslMechanism)) {
+            throw new Error(o.authOpts().saslMechanism + ' mechanism not supported. Supported mechanism are the ones in this list : ' + authMechs);
```

Comment:
Use the same message as the Java client: The selected authentication mechanism '%s' is not among the supported server mechanisms: %s